### ISSUE TITLE
Build crash

### ISSUE THREAD

**Issue by webD97** [minetest/minetest#7074]
##### Issue type
Build issue
 
##### Minetest version
4118e150f8841a507e3fcb546902d1e07150e4a2

##### OS / Hardware
Operating system: Alpine
CPU: arm

##### Summary
Please note that this only happens on arm, but not on amd64. The build uses cmake 3.8.1. For the full log, see https://pastebin.com/JZCUQGjg
```
-- The C compiler identification is GNU 6.3.0
-- The CXX compiler identification is GNU 6.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- *** Will build version 0.5.0-4118e15 ***
-- Found Irrlicht: /usr/lib/libIrrlicht.so  
-- Using GMP provided by system.
-- Found GMP: /usr/lib/libgmp.so  
-- Found JSONCPP: /usr/lib/libjsoncpp.so  
-- Using system JSONCPP library.
-- Found LuaJit: /usr/lib/libluajit-5.1.so  
-- Using LuaJIT provided by system.
-- Found CURL: /usr/lib/libcurl.so  
-- cURL support enabled.
-- GetText disabled.
-- Could NOT find PostgreSQL (missing:  PostgreSQL_LIBRARY PostgreSQL_INCLUDE_DIR PostgreSQL_TYPE_INCLUDE_DIR) 
-- PostgreSQL not found!
-- LevelDB not found!
-- Redis not found!
-- Found SQLite3: /usr/lib/libsqlite3.so  
-- SpatialIndex not found!
-- Found ZLIB: /lib/libz.so (found version "1.2.11") 
-- Looking for clock_gettime in rt
-- Looking for clock_gettime in rt - found
-- Looking for include file endian.h
-- Looking for include file endian.h - found
CMake Error at src/CMakeLists.txt:534 (source_group):
  source_group ROOT: /opt/minetest/src is not a prefix of file:
  /opt/minetest/ban.cpp


CMake Error at src/CMakeLists.txt:535 (source_group):
  source_group ROOT: /opt/minetest/src is not a prefix of file:
  /opt/minetest/ban.cpp


-- Could NOT find Doxygen (missing:  DOXYGEN_EXECUTABLE) 
-- Configuring incomplete, errors occurred!
See also "/opt/minetest/CMakeFiles/CMakeOutput.log".
```

##### Steps to reproduce
This Dockerfile can reproduce this problem, just pass 4118e15 as build arg MINETEST_VERSION. It expects minetest and minetest_game to be in the Docker build context. https://gitlab.com/webD97/minetest-docker/blob/master/Dockerfiles/arm

**Comment by sfan5** [comment]
probably caused by 4118e150f8841a507e3fcb546902d1e07150e4a2
@nOOb3167

**Comment by nOOb3167** [comment]
Investigating

**Comment by nOOb3167** [comment]
Have evidence this failure is caused by a bug in CMake.

https://gitlab.kitware.com/cmake/cmake/issues/16876
https://gitlab.kitware.com/brad.king/cmake/commit/4716f2be83f140deec28fd7c6b945c6fd401a433

It was introduced in CMake 3.8.1 and fixed in later versions.

Here is a minimal reproducing example:

[cmake_issue_repro.zip](https://github.com/minetest/minetest/files/1754618/cmake_issue_repro.zip)

Output of attempting to build with an affected (3.8.1) version:

> C:\Users\Andrej\prog\minetest\ztmpwtf\B381>..\cmake-3.8.1-win32-x86\bin\cmake.exe ..
-- Building for: Visual Studio 15 2017
-- The C compiler identification is MSVC 19.12.25835.0
-- The CXX compiler identification is MSVC 19.12.25835.0
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/
2017/Community/VC/Tools/MSVC/14.12.25827/bin/Hostx86/x86/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/
2017/Community/VC/Tools/MSVC/14.12.25827/bin/Hostx86/x86/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studi
o/2017/Community/VC/Tools/MSVC/14.12.25827/bin/Hostx86/x86/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studi
o/2017/Community/VC/Tools/MSVC/14.12.25827/bin/Hostx86/x86/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at src/CMakeLists.txt:6 (source_group):
  source_group ROOT: C:/Users/Andrej/prog/minetest/ztmpwtf/src is not a
  prefix of file: C:/Users/Andrej/prog/minetest/ztmpwtf/B381/ban.cpp

Output of attempting to build with the current latest version (3.10):

> C:\Users\Andrej\prog\minetest\ztmpwtf\B3_10>"c:\Program Files (x86)\CMake\bin\cmake.exe" ..
-- Building for: Visual Studio 15 2017
-- The C compiler identification is MSVC 19.12.25835.0
-- The CXX compiler identification is MSVC 19.12.25835.0
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/
2017/Community/VC/Tools/MSVC/14.12.25827/bin/Hostx86/x86/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/
2017/Community/VC/Tools/MSVC/14.12.25827/bin/Hostx86/x86/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studi
o/2017/Community/VC/Tools/MSVC/14.12.25827/bin/Hostx86/x86/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studi
o/2017/Community/VC/Tools/MSVC/14.12.25827/bin/Hostx86/x86/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/Andrej/prog/minetest/ztmpwtf/B3_10

Your Dockerfile build for amd64 does _not_ use 3.8.1, which is why it works.
Output (relevant parts) of building the amd64 Dockerfile:

> [root@Fate wtf]# docker build --build-arg MINETEST_VERSION=4118e15 -t wtf .
...
(18/105) Installing cmake (3.9.5-r0)

**Comment by nOOb3167** [comment]
Most Minetest CMakeLists.txt specify source files with a full path (example: ${CMAKE_CURRENT_SOURCE_DIR}/address.cpp).
The exception is src/CMakeLists.txt, which specified source files with a relative path (example: ban.cpp).
Due to the bug, CMake 3.8.1 interprets such paths passed into source_groups relative to the toplevel source directory, instead of the current source directory.

If the relative source file paths in src/CMakeLists.txt are converted to absolute, it would appear this CMake bug can be bypassed. As a bonus all source files would then be specified consistently using the ${CMAKE_CURRENT_SOURCE_DIR}/somefile.cpp notation.

**Comment by SmallJoker** [comment]
Fixed by #7075. Please reopen this issue when it's still persistent.
